extensions/VisualEditor/wikia/modules/ve/ui/dialogs/ve.ui.WikiaInfoboxInsertDialog.js

```diff
@@ -155,26 +139,112 @@ ve.ui.WikiaInfoboxInsertDialog.prototype.insertInfoboxTemplate = function () {
+ * @returns {OO.ui.DecoratedOptionWidget}
+ */
+ve.ui.WikiaInfoboxInsertDialog.prototype.createInfoboxItemOptionWidget = function ( data ) {
+	return new OO.ui.DecoratedOptionWidget( {
+		data: data.title,
+		label:  data.title
+	} );
+};
+
+/**
+ * @desc creates select widget with infobox option widgets
+ * @param {OO.ui.DecoratedOptionWidget[]} items
+ * @returns {OO.ui.OoUiSelectWidget}
+ */
+ve.ui.WikiaInfoboxInsertDialog.prototype.createInfoboxSelectWidget = function ( items ) {
+	var select = new OO.ui.SelectWidget();
+
+	select.addItems( items );
+
+	return select;
+};
+
+/**
+ * @desc creates dialog content
  * @param {Object} data Response data from API
+ * @returns {Promise}
  */
-ve.ui.WikiaInfoboxInsertDialog.prototype.showResults = function ( data ) {
-	var i,
-		items = [],
-		infoboxes = data.query ? data.query.allinfoboxes : [];
-
-	if ( infoboxes.length > 0 ) {
-		for ( i = 0; i < infoboxes.length; i++ ) {
-			items.push(
-				new OO.ui.DecoratedOptionWidget({
-					data: infoboxes[i].title,
-					label:  infoboxes[i].title
-				})
-			);
-		}
-		this.select.addItems( items );
+ve.ui.WikiaInfoboxInsertDialog.prototype.createDialogContent = function ( data ) {
+	var deferred = $.Deferred(),
+		infoboxTemplatesData = data.query ? data.query.allinfoboxes : [],
+		infoboxTemplateOptionWidgets = infoboxTemplatesData.map( this.createInfoboxItemOptionWidget );
+
+	if ( infoboxTemplateOptionWidgets.length > 0 ) {
+		this.select = this.createInfoboxSelectWidget( infoboxTemplateOptionWidgets );
+
+		this.select.connect( this, {
+			select: 'onInfoboxTemplateSelect'
+		} );
+		deferred.resolve( this.select.$element );
+	} else {
+		// creates empty state content
+		this.getUnconvertedInfoboxes()
+			.then( this.createEmptyStateContent.bind( this ) )
+			.then( deferred.resolve )
 	}
+
+	return deferred.promise();
+};
+
+/**
+ * @desc gets list of unconverted infoboxes
+ * @returns {Promise}
+ */
+ve.ui.WikiaInfoboxInsertDialog.prototype.getUnconvertedInfoboxes = function () {
+	var deferred = $.Deferred(),
+		self = this;
+
+	ve.init.target.constructor.static.apiRequest( {
+		action: 'query',
+		list: 'unconvertedinfoboxes'
+	} )
+		.done( function ( data ) {
+			deferred.resolve(
+				self.validateGetUnconvertedInfoboxesResponseData( data ) ? data.query.unconvertedinfoboxes : []
+			);
+		} )
+		.fail( function () {
+			deferred.resolve( [] );
+		} );
+
+	return deferred.promise();
+};
+
+/**
+ * @desc validates get unconverted infoboxes response data
+ * @param {Object} data
+ * @returns {Boolean}
+ */
+ve.ui.WikiaInfoboxInsertDialog.prototype.validateGetUnconvertedInfoboxesResponseData = function ( data ) {
+	return data.query && data.query.unconvertedinfoboxes && data.query.unconvertedinfoboxes.length > 0;
+};
+
+/**
+ * @desc creates empty state dialog content
+ * @param {Array} unconvertedInfoboxes
+ * @returns {Promise}
+ */
+ve.ui.WikiaInfoboxInsertDialog.prototype.createEmptyStateContent = function ( unconvertedInfoboxes ) {
```

Comment:
This is a follow up...
@RafLeszczynski If method doesn't need to return promise because of what it does (e.g. it's not asynchronous) then it shouldn't return promise.
In the previous comment you said that it returns promise because of the particular usage of it (together with another promise)  - from what I remember to .then in jQuery you can pass method which just return data and it should work just fine.